be/src/exec/vectorized/sorting/sort_helper.h

```diff
@@ -0,0 +1,211 @@
+// This file is licensed under the Elastic License 2.0. Copyright 2021-present, StarRocks Limited.
+
+#pragma once
+
+#if defined(__SSE2__)
+#include <emmintrin.h>
+#endif
+
+#include "column/nullable_column.h"
+#include "column/type_traits.h"
+#include "column/vectorized_fwd.h"
+#include "exec/vectorized//sorting//sort_permute.h"
+#include "util/orlp/pdqsort.h"
+
+namespace starrocks::vectorized {
+
+// Comparator for sort
+template <class T>
+struct SorterComparator {
+    static int compare(const T& lhs, const T& rhs) {
+        if (lhs == rhs) {
+            return 0;
+        } else if (lhs < rhs) {
+            return -1;
+        } else {
+            return 1;
+        }
+    }
+};
+
+template <>
+struct SorterComparator<Slice> {
+    static int compare(const Slice& lhs, const Slice& rhs) { return lhs.compare(rhs); }
+};
+
+#ifndef NDEBUG
+template <class PermutationType>
+static std::string dubug_column(const Column* column, const PermutationType& permutation) {
+    std::string res;
+    for (auto p : permutation) {
+        res += fmt::format("{:>5}, ", column->debug_item(p.index_in_chunk));
+    }
+    return res;
+}
+#endif
+
+inline static size_t find_byte(const std::vector<uint8_t>& list, size_t start, uint8_t byte) {
+    const uint8_t* ptr = list.data() + start;
+    const uint8_t* end = list.data() + list.size();
+
+#ifdef __SSE2__
+    // use simd to test 16bytes each time
+    const int SSE2_BYTES = sizeof(__m128i);
+    const __m128i sse_zero = _mm_set1_epi8(byte);
+    const uint8_t* sse_end = ptr + (end - ptr) / SSE2_BYTES * SSE2_BYTES;
+    for (; ptr < sse_end; ptr += SSE2_BYTES) {
+        uint32_t mask = _mm_movemask_epi8(_mm_cmpeq_epi8(_mm_loadu_si128((const __m128i*)(ptr)), sse_zero));
+        int pos = __builtin_ffs(mask);
+        if (pos != 0) {
+            return ptr + pos - 1 - list.data();
+        }
+    }
+#endif
+
+    for (; ptr < end && *ptr != byte; ++ptr) {
+    }
+
+    return ptr - list.data();
+}
+
+// Find position for zero byte, return size of list if not found
+inline static size_t find_zero(const std::vector<uint8_t>& list, size_t start) {
+    return find_byte(list, start, 0);
+}
+
+inline static size_t find_nonzero(const std::vector<uint8_t>& list, size_t start) {
+    return find_byte(list, start, 1);
+}
+
+template <class T, class Container>
+static inline InlinePermutation<T> create_inline_permutation(const SmallPermutation& other,
+                                                             const Container& container) {
+    InlinePermutation<T> inlined(other.size());
+    for (int i = 0; i < other.size(); i++) {
+        int index = other[i].index_in_chunk;
+        inlined[i].index_in_chunk = index;
+        inlined[i].inline_value = container[index];
+    }
+    return inlined;
```

Comment:
is inlined array will be copyed?

Reply:
It will be optimized by C++'s return value optimization. This case is too trivial, so I choose to believe the compiler.